src/python/zquantum/core/estimator.py

```diff
@@ -6,15 +6,41 @@
     expectation_values_to_real,
     concatenate_expectation_values,
 )
-from openfermion import SymbolicOperator, IsingOperator
+from .hamiltonian import group_comeasureable_terms_greedy
+from openfermion import SymbolicOperator, IsingOperator, QubitOperator
 from overrides import overrides
 import logging
 import numpy as np
 import pyquil
-from typing import Tuple, Optional
+from typing import Tuple, Optional, Callable
 
 logger = logging.getLogger(__name__)
 
+DECOMPOSITION_METHODS = {
+    "greedy": group_comeasureable_terms_greedy,
+    "greedy-sorted": lambda qubit_operator: group_comeasureable_terms_greedy(
+        qubit_operator, True
+    ),
+}
+
+
+def get_decomposition_function(decomposition_method: str) -> Callable:
+    """Get a function for Hamiltonian decomposition from it's name.
+
+    Args:
+        decomposition_method: The name of the Hamiltonian decomposition method.
+    
+    Returns:
+        A callable that performs the decomposition.
+    """
+
+    decomposition_function = DECOMPOSITION_METHODS.get(decomposition_method)
+    if decomposition_function is None:
+        raise ValueError(
+            f"Unrecognized decomposition method {decomposition_method}. Allowed values are {list(DECOMPOSITION_METHODS.keys())}"
+        )
+    return decomposition_function
+
```

Comment:
Makes sense! I replaced this with an Enum.

What are your thoughts on the [`BasicEstimator` constructor](https://github.com/zapatacomputing/z-quantum-core/blob/6f0e0db2e9013870439a45ed3050facf1faeb861/src/python/zquantum/core/estimator.py#L98-L113)? I changed it so that it can accept either a callable or a string for `decomposition_method`. When invoking the constructor from python, I imagine using a callable would be preferable for the reasons you mentioned. However when specifying the `decomposition_method` from a workflow, we can only pass basic types like str, int, etc.

Reply:
Actually it seems what I was trying does not work because you cannot access enum values that are functions by name...

Reply:
So it appears there are some problems in having functions as members of an Enum. I tried the workarounds described [here](https://stackoverflow.com/questions/40338652/how-to-define-enum-values-that-are-functions) but they do not let you reference members by name (as in `DecompositionMethods["greedy"]`). Should we just stick to a dictionary, at least for now?

Reply:
I saw your previous implementation but I didn't manage to respond in time. So it's possible to access functions defined in an enum using their name, but you cannot use dictionary-style access for this. Consider the following example:

```python
from enum import Enum

class SomeFunctions(Enum):

    square = lambda x: x ** 2
    cube = lambda x: x ** 3


print(SomeFunctions.square(10)) # Works fine
print(SomeFunctions.cube(2))    # Works fine
print(getattr(SomeFunctions, "cube")(5)) # Also works fine

print(SomeFunctions["cube"]) # KeyError, as cube is considerd a method of SomeFunctions
```

That being said, a dictionary approach fits well here and thinking from the perspective of its usability in workflows the current implementation seems better than what I suggested.